version: '3.8'

services:
  todo:  # Django application service
    build: 
      context: .
      dockerfile: Dockerfile
    volumes: 
      - .:/app
    ports:
      - "8000:8000"      
    depends_on:
      db:
        condition: service_healthy        
    networks:
      - todo_net
    entrypoint: 
      - /bin/sh
      - -c
      - |
        python manage.py makemigrations todo_app &&
        python manage.py migrate &&
        echo "from django.contrib.auth.models import User; User.objects.create_superuser('admin', 'admin@example.com', 'password')" | python manage.py shell &&
        python manage.py runserver 0.0.0.0:8000

  db:  # Postgres database service
    image: postgres:latest
    environment:
      POSTGRES_USER: todouser 
      POSTGRES_PASSWORD: todopassword
      POSTGRES_DB: tododb
    ports:
      - "5432:5432"
    networks:
      - todo_net
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U todouser -d tododb"]
      interval: 10s
      timeout: 5s
      retries: 5

  pgadmin:  # PGAdmin service for interfacing with the database
    image: dpage/pgadmin4:latest
    environment:
      PGADMIN_DEFAULT_EMAIL: admin@todo.com
      PGADMIN_DEFAULT_PASSWORD: adminpassword
    ports:
      - "8080:80"
    networks:
      - todo_net

networks:
  todo_net:
    driver: bridge
